### PR TITLE
Monster Weapon weakness

### PR DESCRIPTION
**Pull request by NRH-AA**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Add ability to make monsters weak to certain weapons. The same as elements.

**Issues addressed:** 
None

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Comment by ranisalt** [comment]
Cool, but the name is terrlble. Maybe `weaponWeakness` instead of just `weapon`, because `getWeaponList` definitely does not recall a weapon _weakness_ list

**Comment by gigastar** [comment]
> Cool, but the name is terrlble. Maybe `weaponWeakness` instead of just `weapon`, because `getWeaponList` definitely does not recall a weapon _weakness_ list

I knew it would come up. I wanted to hear what you guys thought for naming.

File: src/luascript.cpp [hunk 1]
```diff
@@ -3027,6 +3027,9 @@ void LuaScriptInterface::registerFunctions()
 
 	registerMethod("MonsterType", "getElementList", LuaScriptInterface::luaMonsterTypeGetElementList);
 	registerMethod("MonsterType", "addElement", LuaScriptInterface::luaMonsterTypeAddElement);
+	
+	registerMethod("MonsterType", "getWeaponWeaknessList", LuaScriptInterface::luaMonsterTypeGetWeaponWeaknessList);
+	registerMethod("MonsterType", "addWeakness", LuaScriptInterface::luaMonsterTypeAddWeakness);
```
Comment (MillhioreBT) [inline src/luascript.cpp:3032]: `addWeakness`? no is `addWeaponWeakness`?

**Review by MillhioreBT — CHANGES_REQUESTED** [review]

File: src/monster.cpp [hunk 2]
```diff
@@ -626,6 +626,23 @@ BlockType_t Monster::blockHit(Creature* attacker, CombatType_t combatType, int32
 				blockType = BLOCK_ARMOR;
 			}
 		}
+		
+		if (const Player* player = attacker->getPlayer()) {
```
Comment (4mrcn4) [inline src/monster.cpp:630]: inverse that condition and early return blockType

File: src/monster.cpp [hunk 3]
```diff
@@ -626,6 +626,23 @@ BlockType_t Monster::blockHit(Creature* attacker, CombatType_t combatType, int32
...
+			int32_t weaponMod = 0;
+			WeaponType_t weapon = player->getWeaponType();
+			auto it = mType->info.weaponMap.find(weapon);
+			if (it != mType->info.weaponMap.end()) {
+				weaponMod = it->second;
+			}
+
+			if (weaponMod != 0) {
```
Comment (4mrcn4) [inline src/monster.cpp:638]: same here

File: src/monster.cpp [hunk 4]
```diff
@@ -626,6 +626,23 @@ BlockType_t Monster::blockHit(Creature* attacker, CombatType_t combatType, int32
...
+				damage = static_cast<int32_t>(std::round(damage * ((100 - weaponMod) / 100.)));
+				if (damage <= 0) {
```
Comment (4mrcn4) [inline src/monster.cpp:640]: and here

File: src/monster.cpp (hunk 2 above)
Comment (gigastar) [reply src/monster.cpp:630]: Wut?

Comment (4mrcn4) [reply src/monster.cpp:630]: Instead of writing bloated if statement you can check for falseness and `return blocktype` inside

Comment (4mrcn4) [reply src/monster.cpp:630]: so it would be something like
```
if (!(const Player* player = attacker->getPlayer()) {
  return blockType;
}
```

**Pull request closed without merging**